# Capstone Project 2 Data Wrangling

This is the Data Wrangling staep for the following data science problem

Problem statement:

Given past credit card transactions of European cardholders, with the knowledge of the ones that turned out to be fraud, which new transaction can be marked as fraudulent?.

## Data Collection

In [1]:
#Loading python packages
import os
import pandas as pd
import numpy as np

In [2]:
#Checking the current working directory
dirname= os.getcwd()
print(dirname)

C:\Users\saxen\CapstoneProject2


In [3]:
#Changing the working directory to the location of dataset
os.chdir('C:\\MY FOLDER\\SpringBoard\\CapstoneProject2\\')

#Listing the contents of the directory
os.listdir()

['310_23498_compressed_creditcard.csv.zip', 'creditcard.csv', 'Documents']

In [4]:
#Reading the csv file and loading into a dataframe df, then printing the head of the df dataframe 

df = pd.read_csv("creditcard.csv")
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

## Data Organization

Creating some subfolders to keep the project organized

In [5]:
#Printing the folder structure of the current working directory

path = os.getcwd()
print("The path of the current working durectory is: %s" %path)

The path of the current working durectory is: C:\MY FOLDER\SpringBoard\CapstoneProject2


#### Making subfolder 'models' for storing models in future steps

In [6]:
os.mkdir("models")

#### Making subfolder for 'figures' for storing data visualizations in future steps

In [7]:
os.mkdir("figures")

## Data Definition

Reviewing the following

■ Column   names   
■ Data   types   
■ Description   of   the   columns   
■ Counts   and   percents   unique   values   
■ Ranges   of   values 

In [8]:
#Checking the shape of the dataset

df.shape

(284807, 31)

There are 284807 rows and 31 columns. Will check the column names now

In [9]:
#Checking the column names
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

There are only three logical column names namely 'Time', 'Amount' and 'Class'. Remaining columns follow a pattern similar to 'V*' where * ranges from 1 to 28. This data is not easy to understand because of non-logical column names. So, using the metadata can help here.

#### Metadata Information:

Features V1, V2, … V28 are the principal components obtained with PCA, the only features which have not been transformed with PCA are 'Time' and 'Amount'. 
Feature 'Time' contains the seconds elapsed between each transaction and the first transaction in the dataset. 
The feature 'Amount' is the transaction Amount, this feature can be used for example-dependant cost-sensitive learning. 
Feature 'Class' is the response variable and it takes value 1 in case of fraud and 0 otherwise.


Will check the data type of each of these columns now

In [10]:
#Checking data types of columns
df.dtypes

Time      float64
V1        float64
V2        float64
V3        float64
V4        float64
V5        float64
V6        float64
V7        float64
V8        float64
V9        float64
V10       float64
V11       float64
V12       float64
V13       float64
V14       float64
V15       float64
V16       float64
V17       float64
V18       float64
V19       float64
V20       float64
V21       float64
V22       float64
V23       float64
V24       float64
V25       float64
V26       float64
V27       float64
V28       float64
Amount    float64
Class       int64
dtype: object

All  columns starting with 'V' are of type float64. But 'Time' column is of type float64. This is because according to the metadata, it is the elapsed time in seconds between each transaction. So I choose to keep it unchanged. But I would change the name of this column to 'Elapsed Time' for clarity during analysis in the future steps.

In [11]:
df.rename(columns={'Time':'Elapsed Time'}, inplace=True)


#### I will now check the count of non-null values in the columns

In [12]:
#Checking non-null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Elapsed Time  284807 non-null  float64
 1   V1            284807 non-null  float64
 2   V2            284807 non-null  float64
 3   V3            284807 non-null  float64
 4   V4            284807 non-null  float64
 5   V5            284807 non-null  float64
 6   V6            284807 non-null  float64
 7   V7            284807 non-null  float64
 8   V8            284807 non-null  float64
 9   V9            284807 non-null  float64
 10  V10           284807 non-null  float64
 11  V11           284807 non-null  float64
 12  V12           284807 non-null  float64
 13  V13           284807 non-null  float64
 14  V14           284807 non-null  float64
 15  V15           284807 non-null  float64
 16  V16           284807 non-null  float64
 17  V17           284807 non-null  float64
 18  V18 

In [13]:
#Checking the percent of null values
nas=pd.DataFrame(df.isnull().sum().sort_values(ascending=False)/len(df),columns = ['percent'])
pos = nas['percent'] > 0
nas[pos]

Empty DataFrame
Columns: [percent]
Index: []

##### There are no null values in the dataset for any column. So no Imputation is required for this data. 

Further, I will check the number of unique values in each column

In [14]:
#Extracting count of unique values of each column and storing as a list of dictionaries

unique_vals = [print({col:len(df[col].unique())}) for col in df]

{'Elapsed Time': 124592}
{'V1': 275663}
{'V2': 275663}
{'V3': 275663}
{'V4': 275663}
{'V5': 275663}
{'V6': 275663}
{'V7': 275663}
{'V8': 275663}
{'V9': 275663}
{'V10': 275663}
{'V11': 275663}
{'V12': 275663}
{'V13': 275663}
{'V14': 275663}
{'V15': 275663}
{'V16': 275663}
{'V17': 275663}
{'V18': 275663}
{'V19': 275663}
{'V20': 275663}
{'V21': 275663}
{'V22': 275663}
{'V23': 275663}
{'V24': 275663}
{'V25': 275663}
{'V26': 275663}
{'V27': 275663}
{'V28': 275663}
{'Amount': 32767}
{'Class': 2}


The result shows that all columns satrting with 'V' have identical number of unique values. Class is the response variable in this dataset, it has only two unique values. 

Further investigation into 'Class' column will reveal the unique values.

In [15]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

The values in 'Class' column are 0 and 1. But the number of fraudulent transactions are extremely less than the number of non-fraudulent transactions.Therefore, we have an unbalanced dataset.

#### Range of Values

In [16]:
print(df.aggregate([min,max]).T)

                     min            max
Elapsed Time    0.000000  172792.000000
V1            -56.407510       2.454930
V2            -72.715728      22.057729
V3            -48.325589       9.382558
V4             -5.683171      16.875344
V5           -113.743307      34.801666
V6            -26.160506      73.301626
V7            -43.557242     120.589494
V8            -73.216718      20.007208
V9            -13.434066      15.594995
V10           -24.588262      23.745136
V11            -4.797473      12.018913
V12           -18.683715       7.848392
V13            -5.791881       7.126883
V14           -19.214325      10.526766
V15            -4.498945       8.877742
V16           -14.129855      17.315112
V17           -25.162799       9.253526
V18            -9.498746       5.041069
V19            -7.213527       5.591971
V20           -54.497720      39.420904
V21           -34.830382      27.202839
V22           -10.933144      10.503090
V23           -44.807735      22.528412


The Time ranges from 0 to 172792 seconds, which is almost 2 days, so the data analysis will be done on 2 days of transactional data.

Maximum transaction amount is 25691.16 while the least transaction is 0 which looks like a pre-authorisation or zero dollar auth done by company for fraud and risk checking purposes.

The range of values of all columns starting with 'V' have already been transformed using some PCA transformation, so all values range between some negative and positive values. It would be interesting to investigate which among these 'V' columns have the highest variance based on their PCA scores

### Reviewing the summary statistics of the numeric dataset. 

Since the entire datasethas only numeric values, no column needs to be dropped.

In [17]:
#generating summary statistics of df
df.describe()

Elapsed Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  3.919560e-15  5.688174e-16 -8.769071e-15  2.782312e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean  -1.552563e-15  2.010663e-15 -1.694249e-15 -1.927028e-16 -3.137024e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.537294e-16  7.959909e-16  5.367590e-16  4.458112e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   1.453003e-15  1.699104e-15 -3.660161e-16 -1.206049e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

## Duplicate Data Checking

Looking for duplicate rows in the dataset now

In [18]:
#Checking duplicate rows and storing in new dataframe duplicateRowsDF
duplicateRowsDF = df[df.duplicated()]
duplicateRowsDF

Elapsed Time        V1        V2        V3        V4        V5  \
33              26.0 -0.529912  0.873892  1.347247  0.145457  0.414209   
35              26.0 -0.535388  0.865268  1.351076  0.147575  0.433680   
113             74.0  1.038370  0.127486  0.184456  1.109950  0.441699   
114             74.0  1.038370  0.127486  0.184456  1.109950  0.441699   
115             74.0  1.038370  0.127486  0.184456  1.109950  0.441699   
...              ...       ...       ...       ...       ...       ...   
282987      171288.0  1.912550 -0.455240 -1.750654  0.454324  2.089130   
283483      171627.0 -1.464380  1.368119  0.815992 -0.601282 -0.689115   
283485      171627.0 -1.457978  1.378203  0.811515 -0.603760 -0.711883   
284191      172233.0 -2.667936  3.160505 -3.355984  1.007845 -0.377397   
284193      172233.0 -2.691642  3.123168 -3.339407  1.017018 -0.293095   

              V6        V7        V8        V9  ...       V21       V22  \
33      0.100223  0.711206  0.176066 -0.286717  ...  0.046949  0.208105   
35      0.086983  0.693039  0.179742 -0.285642  ...  0.049526  0.206537   
113     0.945283 -0.036715  0.350995  0.118950  ...  0.102520  0.605089   
114     0.945283 -0.036715  0.350995  0.118950  ...  0.102520  0.605089   
115     0.945283 -0.036715  0.350995  0.118950  ...  0.102520  0.605089   
...          ...       ...       ...       ...  ...       ...       ...   
282987  4.160019 -0.881302  1.081750  1.022928  ... -0.524067 -1.337510   
283483 -0.487154 -0.303778  0.884953  0.054065  ...  0.287217  0.947825   
283485 -0.471672 -0.282535  0.880654  0.052808  ...  0.284205  0.949659   
284191 -0.109730 -0.667233  2.309700 -1.639306  ...  0.391483  0.266536   
284193 -0.167054 -0.745886  2.325616 -1.634651  ...  0.402639  0.259746   

             V23       V24       V25       V26       V27       V28  Amount  \
33     -0.185548  0.001031  0.098816 -0.552904 -0.073288  0.023307    6.14   
35     -0.187108  0.000753  0.098117 -0.553471 -0.078306  0.025427    1.77   
113     0.023092 -0.626463  0.479120 -0.166937  0.081247  0.001192    1.18   
114     0.023092 -0.626463  0.479120 -0.166937  0.081247  0.001192    1.18   
115     0.023092 -0.626463  0.479120 -0.166937  0.081247  0.001192    1.18   
...          ...       ...       ...       ...       ...       ...     ...   
282987  0.473943  0.616683 -0.283548 -1.084843  0.073133 -0.036020   11.99   
283483 -0.218773  0.082926  0.044127  0.639270  0.213565  0.119251    6.82   
283485 -0.216949  0.083250  0.044944  0.639933  0.219432  0.116772   11.93   
284191 -0.079853 -0.096395  0.086719 -0.451128 -1.183743 -0.222200   55.66   
284193 -0.086606 -0.097597  0.083693 -0.453584 -1.205466 -0.213020   36.74   

        Class  
33          0  
35          0  
113         0  
114         0  
115         0  
...       ...  
282987      0  
283483      0  
283485      0  
284191      0  
284193      0  

[1081 rows x 31 columns]

There are 1081 rows with duplicate values. It is worth assuming that same transaction at the same time, using the same amount is probably a data redundancy issue while recording the data. 

Checking if any of these duplicate rows have fraudulent transactions

In [19]:
print("Number of duplicate fraudulent transactions are %s" %len(duplicateRowsDF[duplicateRowsDF['Class']==1]))

Number of duplicate fraudulent transactions are 19


There are 19 rows pointing to duplicate fraudulent data as well.

This number of duplicate fraulent transaction data is not too high and therefore dropping them might not impact the dataset much, however I will investigate the impact further. 

So in the next step I will drop all duplicate rows including fraudulent as well as non-fraudulent row, but keep the 1st occurrence of each of these rows to maintain integrity of all unique occurrences

In [20]:
#Dropping duplicate rows from the dataframe and storing into new dataframe dfn

dfn = df.drop_duplicates(keep='first')
print("The shape of the new dataset is now %s" %(str(dfn.shape)))

The shape of the new dataset is now (283726, 31)


#### One final step - Checking impact of dropping duplicates

As per the metadata, there are 492 frauds out of 284,807 transactions which is only 0.172% of all transactions. So it is important to see the change in this percentage after dropping the duplicate rows

In [21]:
percent_R = 100* (len(dfn[dfn['Class']==1])/len(dfn[dfn['Class']==0]))
print("\nIn the reduced dataset there are now %s%% frauds out of all transactions\n" %(np.round(percent_R,3)))
percent_O = 100* (len(df[df['Class']==1])/len(df[df['Class']==0]))

#Putting the shape of original and reduced data into a dataframe for visual comparison

dict ={'Dataset' : ['Original','Reduced'],'Fraud': [len(df[df['Class']==1]), len(dfn[dfn['Class']==1])],'Non-fraud':  [len(df[df['Class']==0]), len(dfn[dfn['Class']==0])],'Total':  [len(df), len(dfn)],'% of Frauds' :[str(np.round(percent_R,3)) + '%',str(np.round(percent_O,3)) + '%'] }

df_final = pd.DataFrame(dict)
df_final


In the reduced dataset there are now 0.167% frauds out of all transactions



Dataset  Fraud  Non-fraud   Total % of Frauds
0  Original    492     284315  284807      0.167%
1   Reduced    473     283253  283726      0.173%

The dataset size now has reduced by almost 0.006%.

Since the 'V' columns are PCA transformed, these repeated transactions might be associated with some valid transactions in the original dataset before PCA tranformation. So, dropping these might make the data inconsistent with the original data.

It could be useful to perform analysis with both datasets separately to see if there is any significant difference in the results when analysing the data by keeping all duplicates vs. dropping duplicates. So, I will maintain the data in both the states and try to do further analysis with both separately.

#### Exporting data to file 

In [22]:
#Exporting the dataframe after dropping duplicates

dfn.to_csv("step2_output_R.csv", index=False)

#Exporting the dataframe without dropping duplicates

df.to_csv("step2_output_O.csv", index=False)
